test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java

```diff
@@ -226,24 +271,40 @@ public void run() throws Throwable
 
             for (TestVersions upgrade : this.upgrade)
             {
-                logger.info("testing upgrade from {} to {}", upgrade.initial.version, upgrade.upgrade.version);
+                logger.info("testing upgrade from {} to {}", upgrade.initial.version, upgrade.upgradeVersions);
                 try (UpgradeableCluster cluster = init(UpgradeableCluster.create(nodeCount, upgrade.initial, configConsumer, builderConsumer)))
                 {
                     setup.run(cluster);
 
-                    for (int n : nodesToUpgrade)
+
+
+                    for (Version nextVersion : upgrade.upgrade)
                     {
-                        cluster.get(n).shutdown().get();
-                        cluster.get(n).setVersion(upgrade.upgrade);
-                        runBeforeNodeRestart.run(cluster, n);
-                        cluster.get(n).startup();
-                        runAfterNodeUpgrade.run(cluster, n);
+                        try
+                        {
+                            for (int n : nodesToUpgrade)
+                            {
+                                cluster.get(n).shutdown().get();
+                                triggerGC();
+                                cluster.get(n).setVersion(nextVersion);
+                                runBeforeNodeRestart.run(cluster, n);
+                                cluster.get(n).startup();
+                                runAfterNodeUpgrade.run(cluster, n);
+                            }
+
+                            runAfterClusterUpgrade.run(cluster);
+
+                            cluster.checkAndResetUncaughtExceptions();
```

Comment:
if logs have a bad issue, trigger early so we don't deal with "there was an error somewhere"... can still happen after the fact... just trying to be nice!